lib/open_telemetry_decorator.ex

```diff
@@ -44,77 +46,106 @@ defmodule OpenTelemetryDecorator do
 
     dynamic_links = Keyword.get(opts, :links, [])
 
-    quote location: :keep do
-      require OpenTelemetry.Tracer, as: Tracer
-      require OpenTelemetry.Span, as: Span
-
-      links =
-        Kernel.binding()
-        |> Enum.into(%{})
-        |> Map.take(unquote(dynamic_links))
-        |> Map.values()
-
-      parent_span = O11y.start_span(unquote(span_name), links: links)
-      new_span = Tracer.current_span_ctx()
-
-      prefix = Attributes.attribute_prefix()
-
-      input_params =
-        Kernel.binding()
-        |> Keyword.delete(:result)
-        |> Keyword.take(unquote(include))
-        |> O11y.set_attributes(namespace: prefix)
-
-      O11y.set_attributes(pid: self())
+    quote location: :keep, generated: true do
+      data =
+        unquote(__MODULE__).prepare(
+          Kernel.binding(),
+          unquote(span_name),
+          unquote(dynamic_links),
+          unquote(include)
+        )
 
       try do
         result = unquote(body)
-
-        # Called functions can mess up Tracer's current span context, so ensure we at least write to ours
-        Tracer.set_current_span(new_span)
-
-        Kernel.binding()
-        |> Keyword.put(:result, result)
-        |> Keyword.merge(input_params)
-        |> Keyword.take(unquote(include))
-        |> O11y.set_attributes(namespace: prefix)
-
+        unquote(__MODULE__).done(result, Kernel.binding(), data)
```

Comment:
oooohhhh, i was trying to figure out how to do this! now that i see it, it's so obvious. thanks!!